tollway/__main__.py

```diff
@@ -13,78 +19,91 @@
     Help,
 )
 from tollway.events import process_duplicate_event, process_late_event
-from tollway.utils import get_date_variation, write_to_file
+from tollway.utils import encode_message, get_date_variation, get_topic, write_to_file
 from tollway.vehicle import create_payload, create_tollway, create_vehicle, get_tollways
 
+app = typer.Typer()
 tollways = get_tollways()
 fake = Faker()
 fake.add_provider(VehicleProvider)
 
 
+@app.command()
 def main(
-    total_events: Annotated[int, typer.Option(help=Help.TOTAL_EVENTS.value)] = 1,
-    event_rate: Annotated[float, typer.Option(help=Help.EVENT_RATE.value)] = 1.0,
+    total_events: Annotated[
+        int, typer.Option(help=Help.TOTAL_EVENTS.value, callback=total_event_callback)
+    ] = 1,
+    event_rate: Annotated[
+        float, typer.Option(help=Help.EVENT_RATE.value, callback=event_rate_callback)
+    ] = 1.0,
     output_file: Annotated[bool, typer.Option(help=Help.OUTPUT_FILE.value)] = False,
     output_filename: Annotated[
-        str, typer.Option(help=Help.OUTPUT_FILENAME.value)
+        str, typer.Option(help=Help.OUTPUT_FILENAME.value, callback=filename_callback)
     ] = "tollway-traffic.json",
-    date_variation: Annotated[
-        bool, typer.Option(help=Help.DATE_VARIATION.value)
-    ] = False,
+    date_variation: Annotated[bool, typer.Option(help=Help.DATE_VARIATION.value)] = False,
     include_late: Annotated[bool, typer.Option(help=Help.INCLUDE_LATE.value)] = False,
     include_duplicate: Annotated[
-        bool, typer.Option(help=Help.INCLUDE_DUPLICATE.value)
+        bool, typer.Option(help=Help.INCLUDE_DUPLICATE.value, callback=behavior_callback)
     ] = False,
+    pubsub: Annotated[bool, typer.Option(help=Help.PUBSUB.value)] = False,
 ):
 
-    tollway = create_tollway(tollways)
     events_log = {
         "past_events_timestamps": [],
         "past_events": [],
         "all_events": [],
     }
 
+    if pubsub:
+        publisher, topic_path = get_topic()
+
     if date_variation:
         include_late = False
         include_duplicate = False
 
     for event_count in range(total_events):
 
         # create new event
+        tollway = create_tollway(tollways=tollways)
         vehicle = create_vehicle(fake=fake)
         payload = create_payload(vehicle=vehicle, tollway=tollway)
 
         # DATE VARIATION
         if date_variation and event_count % DATE_VARIATION_RATE == 0:
-            payload["timestamp"] = get_date_variation(
-                timestamp=payload.get("timestamp")
-            )
-            # push to topic
-            continue
+            payload["timestamp"] = get_date_variation(timestamp=payload.get("timestamp"))
```

Comment:
type hinted as str, so this .get() won't return None I hope?

Reply:
I don't think it should. In the `create_payload` function I'm using `datetime.now...` to define value for the `timestamp` key. Guess what I'm trying to say is I don't think there will be a situation where `timestamp` key doesn't have a value assigned to it unless edge case encountered.

Reply:
yeah I see now: `payload` comes from:

```
        payload = create_payload(vehicle=vehicle, tollway=tollway)
```

Then `create_payload` always sets this key:

```
def create_payload(vehicle: dict, tollway: tuple) -> dict:
    ...
    vehicle["timestamp"] = datetime.now(tz=timezone.utc).strftime(TIMESTAMP_FORMAT)
    return vehicle
```

So maybe instead of .get() just rely on the key being there (because it is) = more explicit:

```
        if date_variation and event_count % DATE_VARIATION_RATE == 0:
            payload["timestamp"] = get_date_variation(timestamp=payload["timestamp"])
```

Reply:
yes, thanks. Spot on. I made this update. mypy was throwing a few more errors regarding my use of .get() so I changed to just the key.

Reply:
Good.